### PR TITLE
Codebreaker-web app

### PR DESCRIPTION
**Pull request by DaniilYusha**

### CODE REVIEW

File: config.ru [hunk 1]
```diff
@@ -0,0 +1,9 @@
+require_relative 'autoload'
+
+use Rack::Reloader
```
Comment (Mikkko) [inline config.ru]: remove in production please

File: .gitignore [hunk 2]
```diff
@@ -1,2 +1,5 @@
 .history
-.DS_Store
\ No newline at end of file
+.DS_Store
+coverage
+app/db/statistics.yml
+app/db/game.yml
```
Comment (exposwingie) [inline .gitignore]: add empty line pls

File: .rubocop.yml [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+
+Style/FrozenStringLiteralComment:
+  Enabled: false
+
+Style/Documentation:
+  Enabled: false
+
+Metrics/BlockLength:
+  Exclude:
+    - 'spec/**/*.rb'
```
Comment (exposwingie) [inline .rubocop.yml]: empty line

File: config.ru (hunk 1 above)
Comment (exposwingie) [inline config.ru]: off in production pls

File: Gemfile [hunk 4]
```diff
@@ -2,18 +2,23 @@
...
-  gem 'rack', '~> 2.2', '>= 2.2.3'
   gem 'fasterer', '~> 0.8.3'
-  gem 'i18n', '~> 1.8', '>= 1.8.5'
   gem 'rubocop', '~> 0.89.1'
   gem 'rubocop-rspec', '~> 1.42'
 end
 
 group :test do
+  gem 'capybara', '~> 3.33'
+  gem 'rack_session_access', '~> 0.2.0'
```
Comment (Mikkko) [inline Gemfile:21]: i think you can remove it

File: spec/spec_helper.rb [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+require 'simplecov'
+
+SimpleCov.start
+SimpleCov.minimum_coverage 90
+
+require_relative '../autoload.rb'
+require 'rack/test'
+require 'rack'
+require 'capybara/rspec'
+require 'rack_session_access/capybara'
```
Comment (Mikkko) [inline spec/spec_helper.rb:10]: i think you can remove this

**Review by Mikkko — COMMENTED** [review]
seems good

File: Gemfile (hunk 4 above)
Comment (DaniilYusha) [reply Gemfile:21]: @Mikkko I can't because I use session thanks to this gem

File: views/game.html.erb [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+                <button type="button" class="btn btn-primary" disabled>
+                  <%= I18n.t :attempts %>
+                  <span class="badge badge-light"><%= game.difficulty.current_attempts %></span>
+                </button>
+              </div>
+
+              <div class="col button">
+                <button type="button" class="btn btn-primary float-right">
+                  <%= I18n.t :hints %>
+                  <span class="badge badge-light"><%= game.difficulty.current_hints %></span>
```
Comment (exposwingie) [inline views/game.html.erb]: Please disable hints button and 'Hints: ' when hints is out

File: spec/spec_helper.rb [hunk 7]
```diff
@@ -0,0 +1,21 @@
+require 'bundler/setup'
+require 'simplecov'
+
+SimpleCov.start
+SimpleCov.minimum_coverage 90
+
+require_relative '../autoload.rb'
+require 'rack/test'
+require 'rack'
```
Comment (exposwingie) [inline spec/spec_helper.rb]: move rack to autoload

File: config.ru (hunk 1 above)
Comment (DaniilYusha) [reply config.ru]: Done

File: spec/spec_helper.rb (hunk 5 above)
Comment (DaniilYusha) [reply spec/spec_helper.rb:10]: Can't remove because I use it now

File: .gitignore (hunk 2 above)
Comment (DaniilYusha) [reply .gitignore]: Done

File: .rubocop.yml (hunk 3 above)
Comment (DaniilYusha) [reply .rubocop.yml]: Done

File: config.ru (hunk 1 above)
Comment (DaniilYusha) [reply config.ru]: Done

File: spec/spec_helper.rb (hunk 7 above)
Comment (DaniilYusha) [reply spec/spec_helper.rb]: Done

File: views/game.html.erb (hunk 6 above)
Comment (DaniilYusha) [reply views/game.html.erb]: @exposwingie I have it below :smile:

File: app/codebreaker_web/helpers/data_management_helper.rb [hunk 8]
```diff
@@ -0,0 +1,15 @@
+module CodebreakerWeb
+  module DataManagementHelper
+    def get_param(param)
+      @request.params[param]
```
Comment (Mikkko) [inline app/codebreaker_web/helpers/data_management_helper.rb]: please remove instance variables from module

File: app/codebreaker_web/helpers/data_management_helper.rb [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+    end
+
+    def session_die
+      @request.session.clear
```
Comment (Mikkko) [inline app/codebreaker_web/helpers/data_management_helper.rb]: as above

File: app/codebreaker_web/helpers/data_management_helper.rb [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+    end
+
+    def session_param(param)
+      @request.session[param]
```
Comment (Mikkko) [inline app/codebreaker_web/helpers/data_management_helper.rb]: as above

File: views/game.html.erb (hunk 6 above)
Comment (exposwingie) [reply views/game.html.erb]: no, do it disabled = true

File: views/menu.html.erb [hunk 11]
```diff
@@ -17,43 +17,43 @@
...
+                <label for="level" class="lead label"><%= I18n.t :difficulty_label %></label>
                 <select class="custom-select" name="level" required="">
-                  <option value="">Choose game level...</option>
-                  <option value="simple">Simple</option>
-                  <option value="middle">Middle</option>
-                  <option value="hard">Hard</option>
+                  <option value=""><%= I18n.t :choose_level %></option>
+                  <option value="easy"><%= I18n.t('levels.easy') %></option>
+                  <option value="medium"><%= I18n.t('levels.medium') %></option>
+                  <option value="hell"><%= I18n.t('levels.hell') %></option>
```
Comment (exposwingie) [inline views/menu.html.erb]: please do as written in the requirements difficulty - attempts amount - hints amount

File: spec/codebreaker_web/web_application_spec.rb [hunk 12]
```diff
@@ -0,0 +1,327 @@
...
+
+    it 'change attempts count by 1' do
+      expect(page).to have_content(difficulty.current_attempts - 1)
+    end
+  end
+
+  context 'when take hint' do
+    before do
+      File.open(TEST_GAME_PATH, 'w') { |f| f.write game.to_yaml }
+      page.set_rack_session(game_path: TEST_GAME_FILE)
```
Comment (Mikkko) [inline spec/codebreaker_web/web_application_spec.rb]: You write it a lot of times. Maybe better to create top level context with 1 before for this? After, you could remove this repeated lines.

File: spec/codebreaker_web/web_application_spec.rb [hunk 13]
```diff
@@ -0,0 +1,327 @@
...
+      visit '/game'
+      click_link I18n.t('links.hint')
+    end
+
+    after { File.delete TEST_GAME_PATH }
```
Comment (Mikkko) [inline spec/codebreaker_web/web_application_spec.rb]: as above. you could write 1 after and don't repeat this

**Review by Mikkko — CHANGES_REQUESTED** [review]
seems good

File: views/statistics.html.erb [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+              <tbody>
+                <% unless statistics.nil? %>
+                  <% statistics.each.with_index(1) do |user, index| %>
+                    <tr>
+                      <th scope="row"><%= index %></th>
+                      <td><%= user[:name] %></td>
+                      <td><%= user[:difficulty] %></td>
+                      <td><%= user[:attempts_total] - user[:attempts_used] %>/<%= user[:attempts_total] %></td>
+                      <td><%= user[:hints_total] - user[:hints_used] %>/<%= user[:hints_total] %></td>
+                      <td><%= user[:date] %></td>
```
Comment (exposwingie) [inline views/statistics.html.erb]: please do as written in the requirements without +0000

File: app/codebreaker_web/web_application.rb [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+      new(env).response.finish
+    end
+
+    def initialize(env)
+      @request = Rack::Request.new(env)
+      @statistics = Codebreaker::StatisticsService.new FILE_PATH
+      @game_path = @request.session[:game_path]
+      @game = load_game(@game_path)
+      @hints = @request.session[:hints].nil? ? [] : @request.session[:hints]
+    end
```
Comment (exposwingie) [inline app/codebreaker_web/web_application.rb:36]: you can move it to private to

File: views/game.html.erb (hunk 6 above)
Comment (DaniilYusha) [reply views/game.html.erb]: @exposwingie ok, understand and done, thanks

File: app/codebreaker_web/web_application.rb (hunk 15 above)
Comment (DaniilYusha) [reply app/codebreaker_web/web_application.rb:36]: done

File: views/statistics.html.erb (hunk 14 above)
Comment (DaniilYusha) [reply views/statistics.html.erb]: Done

File: views/menu.html.erb (hunk 11 above)
Comment (DaniilYusha) [reply views/menu.html.erb]: Done

File: app/codebreaker_web/helpers/data_management_helper.rb (hunk 8 above)
Comment (DaniilYusha) [reply app/codebreaker_web/helpers/data_management_helper.rb]: done

File: app/codebreaker_web/helpers/data_management_helper.rb (hunk 9 above)
Comment (DaniilYusha) [reply app/codebreaker_web/helpers/data_management_helper.rb]: done

File: app/codebreaker_web/helpers/data_management_helper.rb (hunk 10 above)
Comment (DaniilYusha) [reply app/codebreaker_web/helpers/data_management_helper.rb]: done

File: spec/codebreaker_web/web_application_spec.rb (hunk 12 above)
Comment (DaniilYusha) [reply spec/codebreaker_web/web_application_spec.rb]: done

File: spec/codebreaker_web/web_application_spec.rb (hunk 13 above)
Comment (DaniilYusha) [reply spec/codebreaker_web/web_application_spec.rb]: done

**Review by Mikkko — APPROVED** [review]
seems good

**Review by exposwingie — APPROVED** [review]

File: app/codebreaker_web/helpers/data_management_helper.rb [hunk 16]
```diff
@@ -0,0 +1,19 @@
+module CodebreakerWeb
+  module DataManagementHelper
```
Comment (epoberezhny) [inline app/codebreaker_web/helpers/data_management_helper.rb]: Looks more like response helpers

File: app/codebreaker_web/helpers/game_storage_helper.rb [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+    def load_game(filename)
+      filename.nil? || !File.exist?(path_to_game(filename)) ? nil : YAML.load_file(path_to_game(filename))
+    end
+
+    def save_game(game, filename)
+      File.open(path_to_game(filename), 'w') { |f| f.write game.to_yaml }
+    end
+
+    def generate_file_name
+      "game_#{8.times.map { rand(10) }.join}.yml"
```
Comment (epoberezhny) [inline app/codebreaker_web/helpers/game_storage_helper.rb]: FYI you could use `SecureRandom.uuid`

File: app/codebreaker_web/web_application.rb [hunk 18]
```diff
@@ -0,0 +1,127 @@
+module CodebreakerWeb
+  class WebApplication
+    include DataManagementHelper
+    include GameStorageHepler
+
+    attr_reader :statistics
```
Comment (epoberezhny) [inline app/codebreaker_web/web_application.rb]: Why do we need it?

File: config.ru [hunk 19]
```diff
@@ -0,0 +1,10 @@
+require_relative 'autoload'
+
+use Rack::Reloader if ENV.fetch('RACK_ENV', 'development') == 'development'
+use Rack::Static, urls: %w[/bootstrap /jquery], root: 'node_modules'
+use Rack::Static, urls: ['/assets'], root: 'views'
+use Rack::Session::Cookie, key: ENV.fetch('COOKIE_KEY', 'key'),
+                           secret: ENV.fetch('COOKIE_SECRET_KEY', 'secret_key')
+use RackSessionAccess::Middleware if ENV.fetch('RACK_ENV', 'development') == 'development'
```
Comment (epoberezhny) [inline config.ru]: Maybe it would be better to use the same condition as in the autoload?

File: spec/codebreaker_web/web_application_spec.rb [hunk 20]
```diff
@@ -0,0 +1,329 @@
...
+        expect(page).to have_content(difficulty.current_hints - 1)
+      end
+    end
+  end
+
+  context 'when lose the game' do
+    let(:code) { '1234' }
+
+    before do
+      game.difficulty.instance_variable_set(:@current_attempts, 1)
```
Comment (epoberezhny) [inline spec/codebreaker_web/web_application_spec.rb:189]: Try to avoid using of `instance_variable_set` outside of class in the future

**Review by epoberezhny — CHANGES_REQUESTED** [review]

**Comment by epoberezhny** [comment]
Add focus for the input. It will improve user experience

File: app/codebreaker_web/helpers/game_storage_helper.rb (hunk 17 above)
Comment (DaniilYusha) [reply app/codebreaker_web/helpers/game_storage_helper.rb]: done

File: config.ru (hunk 19 above)
Comment (DaniilYusha) [reply config.ru]: ok, done, thanks

File: app/codebreaker_web/helpers/data_management_helper.rb (hunk 16 above)
Comment (DaniilYusha) [reply app/codebreaker_web/helpers/data_management_helper.rb]: renamed, thx :rocket:

File: app/codebreaker_web/web_application.rb (hunk 18 above)
Comment (DaniilYusha) [reply app/codebreaker_web/web_application.rb]: Removed that, thanks

**Review by epoberezhny — APPROVED** [review]

**Pull request merged**